File: android/test-runner/src/main/kotlin/com/cavadvpn/config/ConfigStore.kt
package com.cavadvpn.config

import android.content.SharedPreferences

/**
 * Persists [VpnConfig] to [SharedPreferences].
 */
object ConfigStore {

    const val KEY_SERVER_HOST       = "server_host"
    const val KEY_SERVER_PORT       = "server_port"
    const val KEY_PRIVATE_KEY       = "private_key_hex"
    const val KEY_SERVER_PUBLIC_KEY = "server_public_key_hex"
    const val KEY_DNS_SERVER        = "dns_server"
    const val KEY_MTU               = "mtu"

    fun save(prefs: SharedPreferences, config: VpnConfig) {
        prefs.edit()
            .putString(KEY_SERVER_HOST,       config.serverHost)
            .putInt   (KEY_SERVER_PORT,       config.serverPort)
            .putString(KEY_PRIVATE_KEY,       config.privateKeyHex)
            .putString(KEY_SERVER_PUBLIC_KEY, config.serverPublicKeyHex)
            .putString(KEY_DNS_SERVER,        config.dnsServer)
            .putInt   (KEY_MTU,               config.mtu)
            .apply()
    }

    fun load(prefs: SharedPreferences): VpnConfig? {
        val host = prefs.getString(KEY_SERVER_HOST, null) ?: return null
        return VpnConfig(
            serverHost        = host,
            serverPort        = prefs.getInt   (KEY_SERVER_PORT,       443),
            privateKeyHex     = prefs.getString(KEY_PRIVATE_KEY,       "") ?: "",
            serverPublicKeyHex = prefs.getString(KEY_SERVER_PUBLIC_KEY, "") ?: "",
            dnsServer         = prefs.getString(KEY_DNS_SERVER,        "8.8.8.8") ?: "8.8.8.8",
            mtu               = prefs.getInt   (KEY_MTU,               1400)
        )
    }

    fun clear(prefs: SharedPreferences) {
        prefs.edit()
            .remove(KEY_SERVER_HOST)
            .remove(KEY_SERVER_PORT)
            .remove(KEY_PRIVATE_KEY)
            .remove(KEY_SERVER_PUBLIC_KEY)
            .remove(KEY_DNS_SERVER)
            .remove(KEY_MTU)
            .apply()
    }
}

File: ios/App/Models/ConfigStore.swift
// ConfigStore.swift — Persistent VPN configuration storage (UserDefaults)

import Foundation

/// Persists VPN connection configuration using UserDefaults.
///
/// Keys are namespaced under `com.cavadvpn.` to avoid collisions.
public struct ConfigStore {

    // MARK: - Key constants

    private enum Key {
        static let serverHost       = "com.cavadvpn.serverHost"
        static let serverPort       = "com.cavadvpn.serverPort"
        static let privateKeyHex    = "com.cavadvpn.privateKeyHex"
        static let serverPublicKey  = "com.cavadvpn.serverPublicKey"
        static let knockKey         = "com.cavadvpn.knockKey"
        static let dnsServer        = "com.cavadvpn.dnsServer"
        static let mtu              = "com.cavadvpn.mtu"
    }

    // MARK: - Save

    /// Persist configuration fields to UserDefaults.
    public static func save(
        to defaults: UserDefaults = .standard,
        host: String,
        port: Int,
        privateKeyHex: String?,
        serverPublicKeyHex: String?,
        knockKeyHex: String? = nil,
        dnsServer: String,
        mtu: Int
    ) {
        defaults.set(host,             forKey: Key.serverHost)
        defaults.set(port,             forKey: Key.serverPort)
        defaults.set(privateKeyHex,    forKey: Key.privateKeyHex)
        defaults.set(serverPublicKeyHex, forKey: Key.serverPublicKey)
        defaults.set(knockKeyHex,      forKey: Key.knockKey)
        defaults.set(dnsServer,        forKey: Key.dnsServer)
        defaults.set(mtu,              forKey: Key.mtu)
    }

    // MARK: - Load

    /// Load configuration from UserDefaults; returns nil if the host is missing.
    public static func load(from defaults: UserDefaults = .standard) -> StoredConfig? {
        guard let host = defaults.string(forKey: Key.serverHost), !host.isEmpty else {
            return nil
        }
        let port = defaults.integer(forKey: Key.serverPort)
        let effectivePort = port > 0 ? port : 443
        return StoredConfig(
            host: host,
            port: effectivePort,
            privateKeyHex: defaults.string(forKey: Key.privateKeyHex),
            serverPublicKeyHex: defaults.string(forKey: Key.serverPublicKey),
            knockKeyHex: defaults.string(forKey: Key.knockKey),
            dnsServer: defaults.string(forKey: Key.dnsServer) ?? "1.1.1.1",
            mtu: {
                let v = defaults.integer(forKey: Key.mtu)
                return v > 0 ? v : 1420
            }()
        )
    }

    // MARK: - Clear

    /// Remove all stored configuration from UserDefaults.
    public static func clear(from defaults: UserDefaults = .standard) {
        [Key.serverHost, Key.serverPort, Key.privateKeyHex,
         Key.serverPublicKey, Key.knockKey, Key.dnsServer, Key.mtu].forEach {
            defaults.removeObject(forKey: $0)
        }
    }

    // MARK: - Apply QR

    /// Save configuration from a parsed QR code.
    public static func apply(_ config: ParsedConfig, to defaults: UserDefaults = .standard) {
        save(
            to: defaults,
            host: config.host,
            port: config.port,
            privateKeyHex: config.privateKey,
            serverPublicKeyHex: config.serverKey,
            knockKeyHex: config.knockKey,
            dnsServer: config.dns,
            mtu: config.mtu
        )
    }
}

// MARK: - StoredConfig

/// A fully-loaded configuration retrieved from UserDefaults.
public struct StoredConfig: Equatable {
    public let host: String
    public let port: Int
    public let privateKeyHex: String?
    public let serverPublicKeyHex: String?
    /// Hex-encoded 32-byte port-knock PSK; nil for non-knock servers.
    public let knockKeyHex: String?
    public let dnsServer: String
    public let mtu: Int
}
